modules/index/src/main/java/org/apache/ignite/internal/idx/IndexManagerImpl.java

```diff
@@ -0,0 +1,515 @@
+            throw new IgniteException(new NodeStoppingException());
+        }
+
+        try {
+            return future.join();
+        } catch (CompletionException ex) {
+            throw convertThrowable(ex.getCause());
+        } finally {
+            busyLock.leaveBusy();
+        }
+    }
+
+    /**
+     * Convert to public throwable.
+     *
+     * @param th Throwable.
+     * @return Public throwable.
+     */
+    private RuntimeException convertThrowable(Throwable th) {
+        if (th instanceof RuntimeException) {
+            return (RuntimeException) th;
+        }
+
+        return new IgniteException(th);
+    }
+
+    /**
+     * Gets a cause exception for a client.
+     *
+     * @param t Exception wrapper.
+     * @return A root exception which will be acceptable to throw for public API.
+     */
+    //TODO: IGNITE-16051 Implement exception converter for public API.
+    private @NotNull IgniteException getRootCause(Throwable t) {
+        Throwable ex;
+
+        if (t instanceof CompletionException) {
+            if (t.getCause() instanceof ConfigurationChangeException) {
+                ex = t.getCause().getCause();
+            } else {
+                ex = t.getCause();
+            }
+        } else {
+            ex = t;
+        }
+
+        return ex instanceof IgniteException ? (IgniteException) ex : new IgniteException(ex);
+    }
+
+    private void createIndexLocally(TableImpl tbl, SortedIndexView idxView) {
+        SchemaDescriptor tblSchema = tbl.schemaView().schema();
+
+        Map<String, Column> cols = Stream.concat(
+                Arrays.stream(tblSchema.keyColumns().columns()),
+                Arrays.stream(tblSchema.valueColumns().columns())
+        ).collect(Collectors.toMap(Column::name, Function.identity()));
+
+        List<SortedIndexColumnDescriptor> idxCols = Stream.concat(
+                        idxView.columns().namedListKeys().stream(),
+                        Arrays.stream(tblSchema.keyColumns().columns())
+                                .map(Column::name)
+                )
+                .distinct()
+                .map(colName -> {
+                    IndexColumnView idxCol = idxView.columns().get(colName);
+                    return new SortedIndexColumnDescriptor(
+                            cols.get(colName),
+                            new SortedIndexColumnCollation(idxCol == null || idxCol.asc()));
+                })
+                .collect(Collectors.toList());
+
+        SortedIndexDescriptor idxDesc = new SortedIndexDescriptor(
+                idxView.name(),
+                idxCols,
+                tblSchema.keyColumns().columns()
+        );
+
+        InternalSortedIndexImpl idx = new InternalSortedIndexImpl(
+                idxView.name(),
+                tbl.internalTable().storage().createSortedIndex(idxDesc),
+                tbl
+        );
+
+        tbl.addRowListener(idx);
+
+        idxsByName.put(idx.name(), idx);
```

Comment:
How to deal with duplicates? Is it guaranteed by configuration manager?

What if we have dropped index for table T1, and create index for T2 with the same name.
Is it possible there is a window, when we will have 2 indices with same name due to async index creation/removal neture?